rdr_service/tools/tool_libs/codes_management.py

```diff
@@ -266,21 +297,42 @@ def run(self, skip_file_export_write=False):
                 # Get the data-dictionary and process codes
                 dictionary_json = self.retrieve_data_dictionary(project_api_key)
                 for item_json in dictionary_json:
-                    self.import_data_dictionary_item(session, item_json)
+                    try:
+                        self.import_data_dictionary_item(session, item_json)
+                    except InvalidCodeValue as exc:
+                        invalid_codes_found.append(exc.code_value)
+                    except QuestionOptionsMissing as exc:
+                        questions_missing_options.append(exc.code_value)
 
                 # Don't save anything if codes were unintentionally reused
                 if self.code_reuse_found and not self.args.dry_run:
-                    session.rollback()
                     logger.error('The above codes were already in the RDR database. '
                                  'Please verify with the team creating questionnaires in Redcap that this '
                                  'was intentional, and then re-run the tool with the "--reuse-codes" argument '
                                  'to specify that they should be allowed.')
-                    return 1
-
-            if not self.args.dry_run and not skip_file_export_write:
+                    exit_code = 1
+
+                # Don't save anything if no module code was found
+                if self.module_code is None:
+                    logger.error('No module code found, canceling import')
+                    exit_code = 1
+
+                if invalid_codes_found:
+                    code_values_str = ", ".join([f'"{value}"' for value in invalid_codes_found])
+                    logger.error(f'Invalid code values found: {code_values_str}')
+                    exit_code = 1
+
+                if questions_missing_options:
+                    code_values_str = ", ".join([f'"{value}"' for value in questions_missing_options])
+                    logger.error(f'The following question codes are missing answer options: {code_values_str}')
+                    exit_code = 1
+
+            if exit_code == 1:
```

Comment:
I'm a little confused as to whether an update that needs to be rolled back could have occurred before calling write_export_file?